File: front/lib/api/analytics.ts
import { apiGet } from './base'

export interface CountryFrequencies {
  [country: string]: number
}

export function getCountryFrequencies() {
  return apiGet<CountryFrequencies>('/analytics/frequencies/country')
}

export interface TopVisitor {
  country: string
  visits_count: number
}

/** @deprecated Prefer getTopCountries for public usage UI */
export function getTopVisitors(limit = 5) {
  return apiGet<TopVisitor[]>(`/analytics/top-visitors?limit=${limit}`)
}

export interface UsageSummary {
  unique_users: number
  active_30d: number
  countries: number
  returning_pct: number
  as_of: string
}

export function getUsageSummary() {
  return apiGet<UsageSummary>('/analytics/summary')
}

export interface TopCountry {
  country: string
  unique_users: number
}

export function getTopCountries(limit = 10) {
  return apiGet<TopCountry[]>(`/analytics/top-countries?limit=${limit}`)
}

export interface UsageCapabilityItem {
  id: string
  label: string
  unique_users: number
  request_count?: number
}

export interface UsageCapabilitiesResponse {
  items: UsageCapabilityItem[]
  as_of: string | null
}

export function getUsageCapabilities() {
  return apiGet<UsageCapabilitiesResponse>('/analytics/capabilities')
}

export interface TopEntityRow {
  id: string
  unique_users: number
  label?: string | null
  organism_name?: string | null
  assembly_accession?: string | null
  provider?: string | null
  database?: string | null
  rank?: string | null
}

export interface TopEntitiesResponse {
  top_assemblies: TopEntityRow[]
  top_annotations: TopEntityRow[]
  top_taxons: TopEntityRow[]
  as_of: string | null
}

export function getTopEntities() {
  return apiGet<TopEntitiesResponse>('/analytics/top-entities')
}

File: server/api/analytics.py
from fastapi import APIRouter, Query
from services import analytics_service

router = APIRouter()


@router.get("/analytics/frequencies/country")
async def get_country_frequencies():
    """
    Get frequency counts of unique users by country.

    Returns a dictionary mapping country names to the count of unique users per country.
    Each user is identified by a unique fingerprint (HMAC hash of their IP address).
    """
    return analytics_service.get_country_frequencies()


@router.get("/analytics/top-visitors")
async def get_top_visitors(limit: int = 5):
    """
    Top anonymous visitors by distinct visit days.

    Returns country and visits_count only (no fingerprints or IP data).
    Prefer /analytics/top-countries for public usage UI.
    """
    capped_limit = min(max(limit, 1), 5)
    return analytics_service.get_top_visitors(limit=capped_limit)


@router.get("/analytics/summary")
async def get_usage_summary():
    """
    Public usage hero metrics from UserAnalytics (API-activity users).
    Never returns fingerprints or IPs.
    """
    return analytics_service.get_usage_summary()


@router.get("/analytics/top-countries")
async def get_top_countries(limit: int = Query(default=10, ge=1, le=50)):
    """
    Top countries by unique users (fingerprints), not visit-day counts.
    """
    return analytics_service.get_top_countries(limit=limit)


@router.get("/analytics/capabilities")
async def get_usage_capabilities():
    """
    Product-capability usage from UsageRollup (unique users who touched each bucket).
    Empty items if the daily rollup has not run yet.
    """
    return analytics_service.get_usage_capabilities()


@router.get("/analytics/top-entities")
async def get_top_entities():
    """
    Top-10 opened assemblies, annotations, and taxons by unique users.
    """
    return analytics_service.get_top_entities()
